contributors/design-proposals/loadbalancer-proposal.md

```diff
@@ -0,0 +1,396 @@
+}
+
+type LoadBalancerPhase string
+
+const (
+	// used for LoadBalancers that are not available
+	LoadBalancerPending LoadBalancerPhase = "Pending"
+	// used for LoadBalancers that are working well
+	LoadBalancerPending LoadBalancerPhase = "Running"
+	// used for LoadBalancers that failed to be correctly recycled or deleted after being released from a claim
+	LoadBalanceFailed LoadBalancerPhase = "Failed"
+)
+
+// LoadBalancerIngress represents the status of a load-balancer ingress point:
+// traffic should be sent to an ingress point.
+type LoadBalancerIngress struct {
+	// IP is set for load-balancer ingress points that are IP based
+	// (typically GCE or OpenStack load-balancers)
+	IP string `json:"ip,omitempty"`
+
+	// Hostname is set for load-balancer ingress points that are DNS based
+	// (typically AWS load-balancers)
+	Hostname string `json:"hostname,omitempty"`
+}
+```
+
+### LoadbalancerClaim API
+
+```go
+type LoadbalancerClaim struct {
+	unversioned.TypeMeta `json:",inline"`
+	ObjectMeta           `json:"metadata,omitempty"`
+
+	// Spec defines a loadbalancer owned by the cluster
+	Spec LoadBalancerClaimSpec `json:"spec,omitempty"`
+
+	// Status represents the current information about loadbalancer.
+	Status LoadBalancerClaimStatus `json:"status,omitempty"`
+}
+
+type LoadBalancerClaimSpec struct {
+	// A label query over loadbalancer to consider for binding. This selector is
+	// ignored when LoadBalancerName is set
+	Selector *unversioned.LabelSelector `json:"selector,omitempty"`
+	// Resources represents the minimum resources required
+	Resources ResourceRequirements `json:"resources,omitempty"`
+	// LoadBalancerName is the binding reference to the LoadBalancerName backing this
+	// claim. When set to non-empty value Selector is not evaluated
+	LoadBalancerName string `json:"loadbalancerName,omitempty"`
+}
+
+type LoadBalancerClaimStatus struct {
+}
+```
+
+### Ingress API Changes
+
+Add a new knob "LoadBalancer" in IngressSpec, so that Ingress can "use" or
+"consume" a LoadBalancer, either directly or through LoadBalancerClaim.
+
+```go
+type IngressSpec struct {
+    /*
+    ...
+    */
+
+    LoadBalancer LBSource `json:"loadBalancer"`
+}
+
+type LBSource struct {
+    GCELoadBalancer *GCELoadBalancerSource `json:"gceLoadBalancer,omitempty"`
+	AWSLoadBalancer *AWSLoadBalancerSource `json:"awsLoadBalancer,omitempty"`
+	BareMetalLoadBalancer *BareMetalLoadBalancerSource `json:"bareMetalLoadBalancer,omitempty"`
+	LoadBalancerClaim *LoadBalancerClaimSource `json:"loadBalancerClaim,omitempty"`
+}
+
+type LoadBalancerClaimSource struct {
+	// ClaimName is the name of a LoadBalancerClaim in the same namespace as the ingress using this lb
+	ClaimName string `json:"claimName"`
+}
+```
+
+## Implementation Detail
+
+### loadbalancer
+
+The current loadbalancer (ingress-controller) implementation has some limitations
+when we have multiple loadbalancers:
+
+* we need to deploy multiple ingress-controller pod even on cloud provider, it
```

Comment:
Why do we need to deploy multiple ingress controller pods? Do you mean multiple instances of the same controller (for HA) or different pods for each ingress class? If its the second then, admin can choose to deploy the ingress controllers only for ingress classes they want to enable?

Reply:
No. I want to deploy multiple ingress controller pods. Because L7 lb on cloud provider have limit/quota on some resources, for example, a Aliyun lb have a 0.5G bandwidth upper limit on bandwidth. It's hard to let a single lb to serve all ingress. So I want to deploy multiple ingress controller pod, and schedule ingress to ingress-controller.

Reply:
Proposal should probably clarify distinction between horizontally autoscaling a collection of homogeneous set of ingress pods (likely simpler and more common use-case) vs sharding ingress resources among distinct controllers or groups of pods.

Can you calrify why your case isn't covered by horizontal scaling?

Reply:
Thanks for that great explanation @mqliang! 
I think there is some confusion between the ingress controller pod (that manages LB's for ingresses) and the L7 load balancer pod (that actually balances the user traffic).

I agree that we will need multiple LB pods (due to bandwidh limitation in the Aliyun example) but is still not clear to me why need multiple ingress controller pods. Single aliyun ingress controller can spin up multiple aliyun LB pods?

Reply:
@louiscryan In terms of horizontal scaling, it's complex:
* On GCE/AWS: LB itself is horizontal scalable, basically you can send as many traffic to LB and LB itself will scale up/down accordingly. There is no maximum bandwidth defined for a single LB on aws/gce
* On other cloud provider(like Aliyun): LB is not horizontal scalable, a single LB on Aliyun have a upper limit of 0.5G bandwidth for incoming request. 
* On bare-metal: LB is not horizontal scalable. LB is implemented using  nginx/haproxy and use Keepalived for HA. But Keepalived is a cold backup solution, it's not horizontal scalable(no matter how many LB instance you add, there is only one instance handle the incoming traffic)

Since LB itself is NOT horizontal scalable on some cloud provider(like Aliyun) and bare-metal, we need a mechanism to shard ingress resources among distinct LB.

Reply:
@nikhiljindal You are right: what we actually need is multiple LB, and a mechanism to shard ingress resources among distinct LB. And this is exactly what I want to do in long term.

But current ingress-controller implementation (especially the nginx-ingress-controller) actually put the ingress-controller(list&watch logic) and lb(nginx that handle the incoming traffic) together, in other words, currently ingress-controller and looadbalancer are 1-to-1 relationship, so we can just assign ingress to ingress-controller(this way will also make smallest modification to current ingress-controller implementation).

Reply:
> ingress-controller and looadbalancer are 1-to-1 relationship

Just fyi you can have multiple pod-lbs all listing watching a single ingress resource and adding their IPs to the list of ingress ips. This is how you deploy > 1 replicas (https://github.com/kubernetes/ingress/blob/master/examples/scaling-deployment/nginx/nginx-ingress-deployment.yaml). You need a cloud-lb on top of this to choose between the ips or just  DNS or a smart client.